sway-core/src/type_system/type_mapping.rs

```diff
@@ -1,130 +1,253 @@
-                    name: x.name_ident.clone(),
-                }),
-            )
-        })
-        .collect()
+/// The [TypeMapping] is used to create a mapping between a type that we are
+/// looking for (LHS) and the corresponding type that we want to replace it with
+/// if we find it (RHS).
+pub(crate) struct TypeMapping {
+    mapping: Vec<(TypeId, TypeId)>,
 }
 
-pub(crate) fn create_type_mapping(superset_type: TypeId, subset_type: TypeId) -> TypeMapping {
-    match (look_up_type_id(superset_type), look_up_type_id(subset_type)) {
-        (TypeInfo::Ref(superset_type, _), TypeInfo::Ref(subset_type, _)) => {
-            create_type_mapping(superset_type, subset_type)
-        }
-        (TypeInfo::Ref(superset_type, _), _) => create_type_mapping(superset_type, subset_type),
-        (_, TypeInfo::Ref(subset_type, _)) => create_type_mapping(superset_type, subset_type),
-        (TypeInfo::UnknownGeneric { .. }, _) => {
-            vec![(superset_type, subset_type)]
-        }
-        (
-            TypeInfo::Custom {
-                type_arguments: type_parameters,
-                ..
-            },
-            TypeInfo::Custom { type_arguments, .. },
-        ) => {
-            let type_parameters = type_parameters
-                .unwrap_or_default()
-                .iter()
-                .map(|x| x.type_id)
-                .collect::<Vec<_>>();
-            let type_arguments = type_arguments
-                .unwrap_or_default()
-                .iter()
-                .map(|x| x.type_id)
-                .collect::<Vec<_>>();
-            insert_type_parameters_with_type_arguments(type_parameters, type_arguments)
-        }
-        (
-            TypeInfo::Enum {
-                type_parameters, ..
-            },
-            TypeInfo::Enum {
-                type_parameters: type_arguments,
-                ..
-            },
-        ) => {
-            let type_parameters = type_parameters
-                .iter()
-                .map(|x| x.type_id)
-                .collect::<Vec<_>>();
-            let type_arguments = type_arguments.iter().map(|x| x.type_id).collect::<Vec<_>>();
-            insert_type_parameters_with_type_arguments(type_parameters, type_arguments)
-        }
-        (
-            TypeInfo::Struct {
-                type_parameters, ..
-            },
-            TypeInfo::Struct {
-                type_parameters: type_arguments,
-                ..
+impl TypeMapping {
+    pub(crate) fn from_type_parameters(type_parameters: &[TypeParameter]) -> TypeMapping {
+        let mapping = type_parameters
+            .iter()
+            .map(|x| {
+                (
+                    x.type_id,
+                    insert_type(TypeInfo::UnknownGeneric {
+                        name: x.name_ident.clone(),
+                    }),
+                )
+            })
+            .collect();
+        TypeMapping { mapping }
+    }
+
+    pub(crate) fn from_superset_and_subset(
```

Comment:
Do you mind adding a comment before this method explaining what it does? Specifically, I'm looking for what `superset` and `subset` exactly refer to.

Reply:
Done!